dask/array/core.py

```diff
@@ -592,7 +595,12 @@ def map_blocks(
     original_kwargs = kwargs
 
     if dtype is None and meta is None:
-        dtype = apply_infer_dtype(func, args, original_kwargs, "map_blocks")
+        try:
+            meta = compute_meta(func, dtype, *args, **kwargs)
+        except Exception:
+            dtype = apply_infer_dtype(func, args, original_kwargs, "map_blocks")
+        finally:
+            dtype = apply_infer_dtype(func, args, original_kwargs, "map_blocks")
```

Comment:
I don't like this at all.  I am little confused about `compute_meta.  Apparently, it can return None as part of exception handling:
https://github.com/dask/dask/blob/9e994bad3be012c4ecb03678a88934c4e66cdebb/dask/array/utils.py#L109-L136

I suspect other parts of the codebase rely on this.

Reply:
I think this is par for the course for `_meta`. For example please see this discussion ( https://github.com/dask/dask/pull/4543#issuecomment-491983711 ).

Reply:
cc @pentschev (in case you see a simpler solution)

Reply:
Yes, @jakirkham is right. So what happens in that case is `compute_meta` wasn't able to infer `meta` properly given that particular function and given input(s). In such case we need to return `None` because there's nothing else `compute_meta` can do, at that point it's up to the caller to decide what to do, which can be, for example, just use the `dtype` -- implies a fallback to `np.ndarray`.

However, I do agree that it's strange to have both `except` and `finally` doing exactly the same thing. Maybe it could be rewritten as follows, but please correct me if I failed to consider some case where this wouldn't work:

```python
try:
    meta = compute_meta(func, dtype, *args, **kwargs)
except Exception:
    pass

dtype = apply_infer_dtype(func, args, original_kwargs, "map_blocks")
```

Reply:
Sorry thought for some reason we already made the change. Adding it in PR ( https://github.com/dask/dask/pull/6724 ).